### ISSUE TITLE
for/else is not supported

### ISSUE THREAD

**Issue by sjoerdjob** [pybee/voc#171]
After some experimenting, I ran into a case where the transpiler gives an error.

```
for k in []:
    pass
else:
    print("nobreak")
```

This is as far as I could get to a minimal (non-)working example. Replacing the `print` with `pass` makes it get optimised out by Python already, so the transpiler works.

**Comment by sjoerdjob** [comment]
Never mind. I saw that there's a test-case for it, but it is commented out.  Does it make sense to replace the comment-markers with an `expectFailure` decoration instead? That way it shows up in the test-run as a todo.

**Comment by freakboy3742** [comment]
Thanks for the report, and sorry for the confusion. Yes, it would probably make sense for that test to be marked as an expected failure; I've certainly done that with some other tests. If you want to submit a pull request for this, I'll gladly merge it!

**Comment by Manishearth** [comment]
This was done in https://github.com/pybee/voc/commit/9877c3652206bde636b6d540d70a3eff54e7587f

**Comment by jshipley** [comment]
I've spent some time looking at this, and I'd like to try to fix it. I've learned how it's supposed to work in python bytecode, I just need to learn how to put that into voc and translate it to java bytecode.

Every SETUP_LOOP has a POP_BLOCK at the end. Currently, the POP_BLOCK doesn't do much, but it should be used start the "else" clause or to jump past the "else" clause if the loop was terminated by a BREAK_LOOP.

So far, all I've done is write a few more test cases. When the behavior of SETUP_LOOP, BREAK_LOOP, and POP_BLOCK is correct, then for/else and while/else with or without break statements should all work.

**Comment by sjoerdjob** [comment]
Yes, I have also spent some time looking into this, and I believe your assumptions are correct.

What I did see is that determining the full extent of the loop can be quite complicated, especially for a `while True` loop, which the current approach does not fully recognize (correctly). I would be happy to review some testcases.

**Comment by freakboy3742** [comment]
Fantastic! And congratulations on jumping knee deep into bytecode wrangling :-)

The area where you'll be digging is `voc/python/utils.py`, and in particular the `ForLoop` structure, and the `find_blocks()` and `extract_command()` methods.

When you give VOC a code block, it unpacks all the bytecode into individual instructions, then works backward from the last instruction, finding "sets" of instructions. For example - a `BINARY_ADD` requires two inputs. That means that the instructions preceding the `BINARY_ADD` must place exactly 2 values on the stack. This could be as simple as 2 instructions simply loading a constant, but it might also be a more complex series of instructions that, in total, yields a net 1 instruction added to the stack (e.g., another `BINARY_ADD`, with it's inputs)

So, VOC builds a tree of opcode dependencies, feeding down to the very last opcode (Strictly, it could be a *list* of trees; a new root forms whenever the stack is completely empty). Each node in this tree is called a command; it may have a child nodes that are themselves commands. The role of `extract_commands()` is to turn the list of opcodes into a list of commands.

Layered on top of this are the control structures - for loops, while loops, if statements etc. For example, a while loop is a command that has 2 subcommands - the condition, and the body of the while. These control structures are called Blocks.

So - the job here is to modify `find_blocks()` so that it can identify control structures. A For Loop starts with a SETUP_LOOP followed by FOR_ITER; those opcodes describe the start and end instruction index for the body of the for loop. 

Your job: modify this decomposition logic so that it can:

1. identify the existence of an ELSE block, 
2. store the commands that make up the ELSE block in the ForLoop structure
3. Modify the `transpile()` on ForLoop to output Java opcodes that implement the required logic.

Hope that makes sense. If you need any more pointers, let me know!

**Comment by jshipley** [comment]
I don't want to create a PR yet, but here are the new test cases I have in my fork: https://github.com/jshipley/voc/commit/5c0265727e884f3a42d765ccc0e504cbfc368a42

**Comment by sjoerdjob** [comment]
@jshipley. I saw that in the for-else testcase you don't print anything. This means that not that much is tested.

The while testcases look good.

**Comment by jshipley** [comment]
This problem also affects while loops. On a deeper level, there seems to be some more general issues with nested blocks. A try/except inside a try/except/finally, for example also fails.

To handle more complex cases, the utils.find_blocks function will probably need to use a block stack.

**Comment by Perlence** [comment]
@jshipley Not all while loops have `POP_BLOCK`. For example, the bytecode of your `test_while_forever` lacks one.

**Comment by freakboy3742** [comment]
I'm starting to think that the "decompiling" approach for bytecode might be the wrong approach. As @perlence notes, there is potential for ambiguity (or, at least, extremely difficult decompiling decisions).

Perhaps a better approach might be to use the AST (which knows exactly what the compiled grammar is), and use that to generate equivalent Java bytecode. This would mean rewriting a whole lot of the Python generation code, but it might be a worthwhile activity if we get a cleaner and more complete code generation as a result.

**Comment by sjoerdjob** [comment]
I tend to agree. However if we could get a hold of the bytecode that gets generated before the peephole optimizer passes over it, a lot of the ambiguities would also disappear. However, research into this led to a thread on python-dev without any final decision.

**Comment by jshipley** [comment]
I haven't gotten very deep into planning how to recognize blocks and track them with a block stack. I was reading that every SETUP_LOOP had a POP_BLOCK. If that's not actually true (or if the optimizer makes it untrue), then it may be difficult to accurately identify blocks.

It may indeed be much cleaner to work from the AST. In a way, VOC is already trying to parse the bytecode into an AST--for blocks, at least. I'll spend some time tonight looking at it from the AST direction.

**Comment by jshipley** [comment]
On the other hand:

using the AST rather than the bytecode would eliminate any optimizations that the compiler does, and when I run the code from test_while_forever through `dis.dis()` I do see a POP_BLOCK statement at offset 51.

**Comment by Perlence** [comment]
@jshipley That's very strange. I checked it against Python 3.4.2.

<details>
<summary>Source code</summary>
```
i = 0
while 1:
    print("Loop", i)
    i = i + 1
    if i == 10:
        break
print("Done")
```
</details>

<details>
<summary>Disassembly of while_forever</summary>
```
  1           0 LOAD_CONST               0 (0)
              3 STORE_NAME               0 (i)

  2           6 SETUP_LOOP              42 (to 51)

  3     >>    9 LOAD_NAME                1 (print)
             12 LOAD_CONST               1 ('Loop')
             15 LOAD_NAME                0 (i)
             18 CALL_FUNCTION            2 (2 positional, 0 keyword pair)
             21 POP_TOP

  4          22 LOAD_NAME                0 (i)
             25 LOAD_CONST               2 (1)
             28 BINARY_ADD
             29 STORE_NAME               0 (i)

  5          32 LOAD_NAME                0 (i)
             35 LOAD_CONST               3 (10)
             38 COMPARE_OP               2 (==)
             41 POP_JUMP_IF_FALSE        9

  6          44 BREAK_LOOP
             45 JUMP_ABSOLUTE            9
             48 JUMP_ABSOLUTE            9

  7     >>   51 LOAD_NAME                1 (print)
             54 LOAD_CONST               4 ('Done')
             57 CALL_FUNCTION            1 (1 positional, 0 keyword pair)
             60 POP_TOP
             61 LOAD_CONST               5 (None)
             64 RETURN_VALUE
```
</details>

**Comment by jshipley** [comment]
OK, yeah. I'm running python 3.4.4.

I guess that it will be necessary to decompile everything in 3.4.2 as well when I'm planning/testing everything for this.

**Comment by freakboy3742** [comment]
This should now be a lot easier to implement - the need to reverse engineer structure from bytecode has been replaced by using the AST directly. This makes a lot of structures much easier to handle.

**Comment by eliasdorneles** [comment]
This can be closed, fixed in #357 :tada:
